pkg/logs/decoder/line_generator.go

```diff
@@ -39,11 +54,51 @@ func (l *LineGenerator) Start() {
 	}()
 }
 
-// read reads the input chunks check if match the endline criteria,
+// read reads the input chunks and checks if match the endline criteria,
 // form a line if matches, it also forms a line if the length reaches
 // maxLen limit.
 func (l *LineGenerator) read(chunk *Input) {
-	//TODO
+	i, j := 0, 0
+	n := len(chunk.content)
+	maxj := l.maxLen - l.lineBuf.len()
+
+	for ; j < n; j++ {
+		matchEndLine := l.endLineMatcher.Match(l.lineBuf.contentBytes(), chunk.content, i, j)
+
+		if j == maxj || matchEndLine {
+			l.lineBuf.appendContent(chunk.content[i:j])
+			// when previous line has tailing truncation info, it means
+			// the current line needs leading truncation info.
+			l.lineBuf.lastLeading = l.lineBuf.lastTailing
```

Comment:
I don't think it should be the responsibility of the line buffer to say whether the line is truncated on the left or truncated on the right